### ISSUE TITLE
TLS records are bigger than tune.ssl.maxrecord 

### ISSUE THREAD

**Issue by prueckl** [haproxy/haproxy#1679]
### Detailed Description of the Problem

TLS record fragments are sometimes bigger than configured with tune.ssl.maxrecord.

The first records are in the expected size. Than the record size increases to 16kB.

### Expected Behavior

The Tls records size must not exceed the setting tune.ssl.maxrecord

### Steps to Reproduce the Behavior

1. configure tune.ssl.maxrecord 1419
2. setup wireshark to record the transmitted data
3. download a file (e.g. 200kb)

In my case the backend is apache2. 
The client is an embedded device with little bandwidth (4kB/s) and low RAM (8KB).

On a normal machine the download will succeed. But a small embedded device with low RAM installed will fail to buffer the complete TLS record data before decrypting it.

### Do you have any idea what may have caused this?

I guess that openssl reassembles the data written with SSL_write internally if the buffered data queue is long enough.

### Do you have an idea how to solve the issue?

Setting the SSL_set_max_send_fragment of openssl solved the issue.

```
diff --git a/src/ssl_sock.c b/src/ssl_sock.c
index aedf94477..c903639a9 100644
--- a/src/ssl_sock.c
+++ b/src/ssl_sock.c
@@ -5910,6 +5910,9 @@ static int ssl_sock_handshake(struct connection *conn, unsigned int flag)
 		HA_ATOMIC_INC(&counters_px->reused_sess);
 	}
 
+	if (global_ssl.max_record)
+		SSL_set_max_send_fragment(ctx->ssl, global_ssl.max_record);
+
 	/* The connection is now established at both layers, it's time to leave */
 	conn->flags &= ~(flag | CO_FL_WAIT_L4_CONN | CO_FL_WAIT_L6_CONN);
 	return 1;
```

### What is your configuration?

```haproxy
global
	log /dev/log	local0
	log /dev/log	local1 notice
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
	stats timeout 30s
	user haproxy
	group haproxy
	daemon
	tune.ssl.maxrecord 1419 
#	tune.ssl.maxrecord 2849
#	tune.idletimer 5000

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	# See: https://ssl-config.mozilla.org/#server=haproxy&server-version=2.0.3&config=intermediate
        # ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384


#        ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
        ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets

        ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA
       ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256

defaults
	log	global
	mode	http
	option	httplog
	option	dontlognull
        timeout connect 40000
        timeout client  300s
        timeout server  15m
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http

frontend http_web
	bind 0.0.0.0:443 ssl crt /home/user/mysite.pem
	mode http
	default_backend rgw

backend rgw
	balance roundrobin
	mode http
	server  rgw1 127.0.0.1:80
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.5.6 2022/04/26 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.6.html
Running on: Linux 5.13.0-40-generic #45~20.04.1-Ubuntu SMP Mon Apr 4 09:38:31 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ZLIB=1 USE_NS=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wlallemand** [comment]
Thanks for your report, the suggestion make sense. However the documentation states "Sets the maximum amount of bytes passed to SSL_write() at a time." so we'll need to discuss it with @EmericBr for this change of behavior.

**Comment by wtarreau** [comment]
Hi,

I understand why you need this but this would break a number of setups. The `maxrecord` setting was only introduced with page load time in mind, and over the last 8 years since commit 7bed945be ("OPTIM: ssl: implement dynamic record size adjustment") it has been dynamic in that it's only enforced on non-streaming data but never on streaming data. The idea is that the first few fragments of a page are made of small chunks that the browser can start to decrypt and parse, and that the rest or large objects are sent as large records.

By forcing all records to be small, we would be back to the initial patch that ruined performance and that would force many users to disable `maxrecord`.

Instead I think we should have a different setting, like `hard-maxrecord` to set the upper bound that may never be passed. This way you could adjust this one for your low-footprint machines regardless of the ability to automatically adapt to the traffic.

I'm re-categorizing this as a feature request because it currently works as designed (and admittedly the doc could be improved), but that's something easy enough to do that it could be merged quickly and even backported.

Since you showed that you're able to patch the code, please could you handle this ? Have a look at the parser for "maxrecord" in cfgparse-ssl.c, add one field to the `global_ssl` struct in ssl_sock-t.h, and just add a test like this and it should do the job:
```
diff --git a/src/ssl_sock.c b/src/ssl_sock.c
index 19e41fd87..01f76d301 100644
--- a/src/ssl_sock.c
+++ b/src/ssl_sock.c
@@ -6568,6 +6568,9 @@ static size_t ssl_sock_from_buf(struct connection *conn, void *xprt_ctx, co>
                if (try > count)
                        try = count;
 
+               if (global_ssl.hard_max_record && try > global_ssl.hard_max_record)
+                       try = global_ssl.hard_max_record;
+
                if (!(flags & CO_SFL_STREAMER) &&
                    !(ctx->xprt_st & SSL_SOCK_SEND_UNLIMITED) &&
                    global_ssl.max_record && try > global_ssl.max_record) {
```

Please update the doc to mention the new keyword and that's definitely something we can merge and backport!

**Comment by prueckl** [comment]
patch sent to 'haproxy@formilux.org'

**Comment by wtarreau** [comment]
Merged, marked for backport to 2.5 so that we don't forget.
Thanks!

**Comment by capflam** [comment]
backported to 2.5 now
